### PR TITLE
Revert "Revert "Simplified invokes: SDK-gen and program-gen implementation for dotnet and nodejs""

### PR DESCRIPTION
**Pull request by Zaid-Ajaj**
Reverts pulumi/pulumi#11701 and fixes the problem which caused the issue in the first place:
 - Binding `Outputs` field from `FunctionSpec` into `Function.ReturnType` when available
 - Adding a test that verifies the above works as expected

### CODE REVIEW

**Comment by iwahbe** [comment]
@Zaid-Ajaj Can you highlight the difference between https://github.com/pulumi/pulumi/pull/11418 and https://github.com/pulumi/pulumi/pull/11753? Are the only changes in `schema.go`?

File: pkg/codegen/schema/bind.go [hunk 1]
```diff
@@ -1476,40 +1476,125 @@ func (t *types) bindFunctionDef(token string) (*Function, hcl.Diagnostics, error
...
+			if err != nil {
+				return nil, diags, fmt.Errorf("error binding outputs for function %v: %w", token, err)
+			}
+			returnType = out
+		} else {
+			// Setting `spec.ReturnType` to a value without setting either `TypeSpec` or `ObjectTypeSpec`
+			// indicates a logical bug in our marshaling code.
+			return nil, diags, fmt.Errorf("error binding outputs for function %v: invalid return type", token)
+		}
+	} else if spec.Outputs != nil {
```
Comment (Zaid-Ajaj) [inline pkg/codegen/schema/bind.go:1574]: @iwahbe This check here is the main difference. This is the new code block
```go
if spec.ReturnType != nil && spec.Outputs == nil {
  // bind from return type
} else if spec.Outputs != nil { 
  // bind from spec.Outputs
}
```
whereas before it was only:
```go
if spec.ReturnType != nil  {
  // bind from return type
}
```
Which caused the issue that @danielrbradley had where `spec.Outputs` was specified by code, not through JSON/YAML parsing and didn't bind correctly. 

I've also added a test to show that this actually works: `Function.ReturnType` being bound when `FunctionSpec.Outputs` is specified.

File: pkg/codegen/schema/schema.go [hunk 2]
```diff
@@ -1510,15 +1546,66 @@ type ResourceSpec struct {
...
 	Inputs *ObjectTypeSpec `json:"inputs,omitempty" yaml:"inputs,omitempty"`
+	// Determines whether the input bag should be treated as a single argument or as multiple arguments.
+	// When MultiArgumentInputs is non-empty, it must match up 1:1 with the property names in of the Inputs object.
+	// The order in which the properties are listed in MultiArgumentInputs determines the order in which the
+	// arguments are passed to the function.
+	MultiArgumentInputs []string `json:"multiArgumentInputs,omitempty" yaml:"multiArgumentInputs,omitempty"`
 	// Outputs is the bag of output values for the function, if any.
-	Outputs *ObjectTypeSpec `json:"outputs,omitempty" yaml:"outputs,omitempty"`
-	// DeprecationMessage indicates whether or not the function is deprecated.
+	Outputs    *ObjectTypeSpec `json:"outputs,omitempty" yaml:"outputs,omitempty"`
```
Comment (danielrbradley) [inline pkg/codegen/schema/schema.go:1606]: Deprecation notice - explain why not to specify both `Outputs` and `ReturnType` - i.e. they get written to the same field.

File: pkg/codegen/schema/schema.go [hunk 3]
```diff
@@ -1527,6 +1614,152 @@ type FunctionSpec struct {
...
+// which otherwise isn't possible when both are retrieved from the same JSON field
+func (funcSpec *FunctionSpec) UnmarshalYAML(node *yaml.Node) error {
+	var data map[string]RawMessage
+	if err := node.Decode(&data); err != nil {
+		return err
+	}
+	return unmarshalFunctionSpec(funcSpec, data)
+}
+
+func (funcSpec FunctionSpec) marshalFunctionSpec() map[string]interface{} {
```
Comment (danielrbradley) [inline pkg/codegen/schema/schema.go:1710]: Raise an error if both `Outputs` and `ReturnType` are specified.

**Review by danielrbradley — CHANGES_REQUESTED** [review]

File: pkg/codegen/schema/schema.go (hunk 3 above)
Comment (Zaid-Ajaj) [reply pkg/codegen/schema/schema.go]: Done ✅

File: pkg/codegen/schema/schema.go [hunk 4]
```diff
@@ -1510,15 +1546,66 @@ type ResourceSpec struct {
...
+	// Determines whether the input bag should be treated as a single argument or as multiple arguments.
+	// When MultiArgumentInputs is non-empty, it must match up 1:1 with the property names in of the Inputs object.
+	// The order in which the properties are listed in MultiArgumentInputs determines the order in which the
+	// arguments are passed to the function.
+	MultiArgumentInputs []string `json:"multiArgumentInputs,omitempty" yaml:"multiArgumentInputs,omitempty"`
 	// Outputs is the bag of output values for the function, if any.
-	Outputs *ObjectTypeSpec `json:"outputs,omitempty" yaml:"outputs,omitempty"`
-	// DeprecationMessage indicates whether or not the function is deprecated.
+	Outputs    *ObjectTypeSpec `json:"outputs,omitempty" yaml:"outputs,omitempty"`
+	ReturnType *ReturnTypeSpec
```
[the remark this answers is not in the document]
Comment (Zaid-Ajaj) [reply pkg/codegen/schema/schema.go:1610]: Done ✅

File: pkg/codegen/schema/bind.go [hunk 5]
```diff
@@ -1476,40 +1476,125 @@ func (t *types) bindFunctionDef(token string) (*Function, hcl.Diagnostics, error
...
-	if spec.Outputs != nil {
+
+	language := make(map[string]interface{})
+	for name, raw := range spec.Language {
+		language[name] = json.RawMessage(raw)
+	}
+
+	var inlineObjectAsReturnType bool
+	var returnType Type
+	if spec.ReturnType != nil && spec.Outputs == nil {
```
[the remark this answers is not in the document]
Comment (Zaid-Ajaj) [reply pkg/codegen/schema/bind.go:1554]: Done ✅

File: pkg/codegen/schema/pulumi.json [hunk 6]
```diff
@@ -521,6 +521,13 @@
...
+                "multiArgumentInputs": {
+                    "description": "A list of parameter names that determines whether the input bag should be treated as a single argument or as multiple arguments. The list corresponds to the order in which the parameters should be passed to the function.",
+                    "type": "array",
+                    "items": {
+                        "type": "string"
+                    }
+                },
                 "outputs": {
                     "description": "The bag of output values for the function, if any.",
                     "$ref": "#/$defs/objectTypeSpec"
```
[the remark this answers is not in the document]
Comment (Zaid-Ajaj) [reply pkg/codegen/schema/pulumi.json]: Done ✅

**Review by danielrbradley — APPROVED** [review]

**Comment by Zaid-Ajaj** [comment]
bors merge

**Comment by Zaid-Ajaj** [comment]
bors merge

**Pull request merged**